zebROS_ws/src/tf_object_detection/src/object_detection.cpp

```diff
@@ -26,30 +29,288 @@ void camera_info_callback(const sensor_msgs::CameraInfoConstPtr &info)
+	depth.copyTo(depthDifferentFormat);
+	cv::normalize(depthDifferentFormat, depthDifferentFormat, 0, 128, cv::NORM_MINMAX); // 0-128 because outliers will be 255 (and we want a clear background)
+	depthDifferentFormat.setTo(255, zeros); // set zeros (no depth) to 255
+	depthDifferentFormat.convertTo(depthDifferentFormat, CV_8UC1, 1);
+
+	// Find the median of the image
+	float median = findMedianOfMat(depthDifferentFormat);
+
+	// Create a cv::Mat for thresholding output
+	cv::Mat threshOutput;
+
+	if (adaptive) {
+		// use adaptive thresholding to convert image to black and white for finding
+		// contours
+		int blockSize = depthDifferentFormat.size().area() / 25;
+		blockSize = blockSize > 1 ? (blockSize % 2 == 0 ? blockSize + 1 : blockSize) : 5; // block size must be at least 3 and odd
+		cv::adaptiveThreshold(depthDifferentFormat, threshOutput, 1, cv::ADAPTIVE_THRESH_MEAN_C, cv::THRESH_BINARY_INV, blockSize, 0);
+	} else {
+		// thresholding using median
+		cv::Mat threshOutput;
+		cv::threshold(depthDifferentFormat, threshOutput, median, 1, cv::THRESH_BINARY_INV);
+	}
+
+	// find contours
+	std::vector<std::vector<cv::Point>> contours;
+	std::vector<cv::Vec4i> hierarchy;
+	cv::findContours(threshOutput, contours, hierarchy, cv::RETR_TREE, cv::CHAIN_APPROX_SIMPLE);
+
+	// create a mask
+	cv::Mat mask = cv::Mat::zeros(threshOutput.size(), CV_8UC1);
+	int largestContourIndex;
+	float largestArea = 0;
+	for (size_t i = 0; i < contours.size(); i++) {
+		std::vector<cv::Point> contour = contours[(int)i];
+		cv::Rect rect = cv::boundingRect(contour);
+		if (rect.area() > largestArea) {
+			largestArea = rect.area();
+			largestContourIndex = (int)i;
+		}
+	}
+	// draw the largest contour onto the mask
+	cv::Scalar color = cv::Scalar(255);
+	cv::drawContours(mask, contours, largestContourIndex, color, -1, cv::LINE_8, hierarchy, 0);
+
+	// make a new image for the original depth cut out by the mask, and fill it with 999
+	// 999 = ignore value, see line 168
+	cv::Mat masked = cv::Mat::ones(depth.size(), depth.type()) * 999;
+	depth.copyTo(masked, mask);
+
+	// if debug is enabled,
+	if (debug) {
+		// show the masked image
+		cv::Mat destination = cv::Mat::zeros(depth.size(), depth.type());
+		depth.copyTo(destination, mask);
+
+		cv::normalize(destination, destination, 0, 255, cv::NORM_MINMAX);
+		destination.convertTo(destination, CV_8UC1);
+		cv::imshow("Masked", destination);
+		cv::waitKey(0);
+	}
+
+	// copy matrix data to a vector
+	std::vector<float> vec;
+	masked = masked.reshape(0, 1);
+	masked.copyTo(vec);
+	// remove 999 values from the mask
+	vec.erase(std::remove(vec.begin(), vec.end(), 999), vec.end());
+	// and 0 values (no depth)
+	vec.erase(std::remove(vec.begin(), vec.end(), 0), vec.end());
+	// sort vector
+	std::sort(vec.begin(), vec.end());
+
+	return (vec.size() != 0 ? vec[0] : -1); // return lowest value. Ideally, we'd want to throw out the bottom
```

Comment:
I'm curious about this part.  In real life, the data will have some noise in it, so I'd have guessed that taking the 95%, 98%, 99% percentile lowest value would be safer than using the absolute lowest value.  Is that not true with the test data you've been able to generate?

Reply:
Depending on what those tests show, it might be useful to make the %tile value returned configurable, e.g. 100% (1.0) would mean return the absolute lowest value, .99 would skip the 1% smallest values assuming they're noise, etc.

Reply:
I definitely agree -- with a few of the test cases (like [this one](https://github.com/FRC900/2020Offseason/blob/depth_calculation/zebROS_ws/src/tf_object_detection/test_bitmaps/power_cell.png)) the actual depth was only a few values in the center, but actual data probably wouldn't have that issue.